### ISSUE TITLE
Peer to Peer protocol

### ISSUE THREAD

**Issue by ps-balucan** [digitalsirkeith/coe151#2]
### Login Phase
1. Upon starting the program, the user will be prompted for a username. The program will then send a broadcast to the broadcast IP and will then wait for discovery responses at own IP. The discovery responses will contain the usernames of already active peers (have approved usernames and all)
2. From the discovery responses, the program will be able to collate a list of active usernames. If the desired username is in that list, try another username and resend broadcast to check for duplicates again.
3. After step2, the username inputted by the user is considered as valid.
4. With a valid username, the program will continuously monitor the broadcast ip and send discovery responses to any new broadcasts sent to the broadcast ip.
Relevant formats:
Sent to the broadcast IP:
`{"mtp" : "Discovery" , "data": { "ip" : "127.0.0.1" , "port" : 15151 } }`
Sent to process-specific IP/port obtained from the broadcast above
`{"mtp" : "DiscoveryResponse" , "data": { "name" : "testuser" , "ip" : "127.0.0.1" , "port" : 15151 } }`
### Finding peers
1. Similar to the login phase, the program will send out a broadcast to the broadcast ip
2. The number of discovery responses signify the number of active peers in the network. This list will be collated and their usernames will be displayed.

### Chat Phase
1. The program will be monitoring its own IP/port advertised and display any messages sent to that. 
2. When the user wants to send to another user, he chooses from the peers obtained from the previous part. The IP and ports collected from the discovery responses should be sufficient to open a socket and send a UDP message directly to the other peer (In which that peer will be monitoring his own IP/port and print out the message received.)
Sent to process-specific IP/port obtained from the broadcast above
`{"mtp" : "SendChat" , "data": { "username" : "testuser" ,"message": "Hello World!"  } }`

**Comment by digitalsirkeith** [comment]
Some suggestions:
I think there should be three running phases/state: _Discovery_, _Handshake_, _Chat_.
## Discovery Phase
Discovery phase would be exactly as suggested.
## Handshake Phase
### Handshake
Handshake phase is where a peer now attempts to join the network of other peers. The incoming peer __broadcasts__ a _Handshake_ message. i.e. it sends the following json message:
`{"mtp": "Handshake", "data": {"ip": "192.168.0.1", "port": "15151", "username": "belay"}}`
### HandshakeResponse
All peers within the subnet will reply with a _HandshakeResponse_ through the ip received. i.e.
`{"mtp": "HandshakeResponse", "data": {}, "status": "OK"}`
`{"mtp": "HandshakeResponse", "data": {}, "status": "DuplicateError"}`
It also worth noting that this response is __not__ sent through the broadcast IP.
## Chat Phase
I think we should differentiate target messages (or whispers) from broadcasted chat messages.
### SendChat
This is general chat that is sent through the __broadcast ip__.
`{"mtp" : "SendChat" , "data": {"message": "Hello World!" } }`
### WhisperChat
This is whisper chat that is sent through the direct ip in the peer's local copy of other peer's addresses.
`{"mtp" : "WhisperChat" , "data": {"message": "Hello World!" } }`

## TLDR
- I think username should be removed from SendChat and WhisperChat as the peer should already know the username of the source from the ip.
- I think the peer should collect first a list before letting the user choose. In this way, the user knows what the unavailable usernames are.
- HandshakePhase should be done because we are now trying to introduce/identify ourselves to other peers in the subnet.

**Comment by digitalsirkeith** [comment]
I made some test cases. For peer-peer communication using UDP, **you do not necessarily need a client socket**.  Server to Server messages are possible since there is no explicit connection. This is probably because server socket does not know if the other socket is a server or client, it is only aware that there are segments being sent. I recommend everyone to just use a server socket to keep things much simpler.

**Comment by digitalsirkeith** [comment]
Apparently, we have to set a time-limit between switching from discovery phase to handshake phase. Setting different values for each peer will not necessarily affect the discovery process as long as we set a time limit long enough for everyone in the network to respond. Like this message if you wish to standardize this time limit, unlike otherwise. I personally think one second or more would suffice and it not need be the same for everyone else.

**Comment by digitalsirkeith** [comment]
Implemented na the three phases. HandshakePhase and ChatPhase can be the same.

**Comment by digitalsirkeith** [comment]
Messages that are only applicable to Peer-Peer communication:
- Discovery
- DiscoveryResponse
- Handshake
- HandshakeResponse
- SendChat
- Whisper
- Disconnect
- SetUsername
- SetUsernameResponse

OnlineList and LocalTime are not applicable to peer-peer because these are information that are in the local list each peer has. Admins are not applicable because the privileges of these are given to all users by default.

**Comment by digitalsirkeith** [comment]
I'm actually not sure if we need 4 custom features in our peer-peer communication. MuteUser, KickUser are essentially the same as disconnecting from them. UnmuteUser is the same as re-sending the Handshake message.

**Comment by digitalsirkeith** [comment]
### SetUsername
`{"mtp": "SetUsername", "data": {"name": "loid"}}`

### SetUsernameResponse
`{"mtp": "SetUsernameResponse", "data": {}, "status": "OK"}`
`{"mtp": "SetUsernameResponse", "data": {}, "status": "DuplicateError"}`

**Comment by digitalsirkeith** [comment]
When implementing this protocol/application, I strongly recommend everyone to use event-driven programming as its core concept. Procedural programming will be very difficult to do in this scenario.

**Comment by claudynetrixie** [comment]
> Messages that are only applicable to Peer-Peer communication:
> 
> * Discovery
> * DiscoveryResponse
> * Handshake
> * HandshakeResponse
> * SendChat
> * Whisper
> * Disconnect
> * SetUsername
> * SetUsernameResponse
> 
> OnlineList and LocalTime are not applicable to peer-peer because these are information that are in the local list each peer has. Admins are not applicable because the privileges of these are given to all users by default.

Clarifications:
1. If the user is currently in the Handshake phase (username not yet validated), will it be able to receive broadcasted messages? 
2. Will the disconnect request be broadcasted to all the users in the network even if they're still in the Handshake phase?

**Comment by digitalsirkeith** [comment]
> > Messages that are only applicable to Peer-Peer communication:
> > 
> > * Discovery
> > * DiscoveryResponse
> > * Handshake
> > * HandshakeResponse
> > * SendChat
> > * Whisper
> > * Disconnect
> > * SetUsername
> > * SetUsernameResponse
> > 
> > OnlineList and LocalTime are not applicable to peer-peer because these are information that are in the local list each peer has. Admins are not applicable because the privileges of these are given to all users by default.
> 
> Clarifications:
> 
> 1. If the user is currently in the Handshake phase (username not yet validated), will it be able to receive broadcasted messages?
> 2. Will the disconnect request be broadcasted to all the users in the network even if they're still in the Handshake phase?

1. In my implementation, after sending the Handshake message, it should be able to read all broadcasted messages. However, it will only process messages that are:
- sent by one of its peers in its list. If a message received is from a unrecognized source, it will filter it out.
- HandshakeResponse message

2. Yes, a peer should ignore a broadcasted message if it is from an unrecognized source.

Note: Unrecognized source refers to peers that have not sent a Handshake or HandshakeResponse to me (i.e. peers not in my local list)

**Comment by digitalsirkeith** [comment]
Let's start drafting the RFC document later.

**Comment by digitalsirkeith** [comment]
Peer-Peer Protocol Finished
